### PR TITLE
implement async api to use in offline mode

### PR DESCRIPTION
**Pull request by vpodlesnyak**
added firestore async api

### CODE REVIEW

File: firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt [hunk 1]
```diff
@@ -66,6 +72,8 @@ actual class FirebaseFirestore(val android: com.google.firebase.firestore.Fireba
 
 actual class WriteBatch(val android: com.google.firebase.firestore.WriteBatch) {
 
+    actual val async = Async(android)
```
Comment (vpodlesnyak) [inline firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:75]: in order not to break existing API each class that supports async operations provides access to an `async` field

File: firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/DeferredExtensions.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package dev.gitlive.firebase.firestore
+
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.Deferred
+
+@PublishedApi
+internal fun <T, R> Deferred<T>.convert(converter: (T) -> R): Deferred<R> {
```
Comment (vpodlesnyak) [inline firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/DeferredExtensions.kt:7]: is there a better way to convert a result of a deferred?

File: firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt [hunk 3]
```diff
@@ -21,6 +23,10 @@ actual val Firebase.firestore get() =
 actual fun Firebase.firestore(app: FirebaseApp) =
     FirebaseFirestore(com.google.firebase.firestore.FirebaseFirestore.getInstance(app.android))
 
+@Suppress("DeferredIsResult")
+@PublishedApi
+internal fun Task<Void>.asUnitDeferred(): Deferred<Unit> = asDeferred().convert { }
```
[the remark this answers is not in the document]
Comment (thoutbeckers) [reply firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:28]: Yeah canceling the actual Task would be dangerous and would at least need to be consistent with iOS and JS.

File: firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/DeferredExtensions.kt (hunk 2 above)
Comment (thoutbeckers) [reply firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/DeferredExtensions.kt:7]: Can't think of anything.

**Review by thoutbeckers — APPROVED** [review]
I like this solution, let's hope it can also go upstream.

An alternate would be to just wrap the regular methods in `async { } ` again, but I prefer this approach.

**Comment by vpodlesnyak** [comment]
@thoutbeckers I was thinking on the `async { }` wrappers in order not to update the api, however I couldn't think of an elegant way to release a reference to the data that is being saved

**Pull request merged**